File: clients/typescript/src/connection.ts
import net from "node:net";
import tls from "node:tls";
import { once } from "node:events";

import { ConnectionError, fromStatus } from "./errors.ts";
import { encodeHeader, HEADER_LEN, PROTO_VERSION, Status, Cmd } from "./protocol.ts";

/** A decoded response: a status byte and (possibly empty) payload. */
export interface Response {
  status: number;
  body: Buffer;
}

/** TLS options passed through to `tls.connect`. `true` enables system defaults. */
export type TlsOption = boolean | tls.ConnectionOptions;

interface Pending {
  resolve: (r: Response) => void;
  reject: (err: Error) => void;
  timer: NodeJS.Timeout;
}

interface WriteTask {
  frame: Buffer;
  reject: (err: Error) => void;
}

/**
 * A single TCP connection to a ZydecoDB server. Safe for concurrent use:
 * requests are pipelined. The writer queue respects OS backpressure (drain),
 * and the reader loop continuously consumes frames to resolve promises in FIFO order.
 */
export class Connection {
  private socket: net.Socket | null = null;
  private inFlight: Pending[] = [];
  private writeQueue: WriteTask[] = [];
  private writing = false;
  private dead = false;
  lastUsed = 0;

  private readonly host: string;
  private readonly port: number;
  private readonly timeoutMs: number;
  private readonly apiKey: string | null;
  private readonly tlsOpt: TlsOption | null;

  constructor(
    host: string,
    port: number,
    timeoutMs: number,
    apiKey: string | null,
    tlsOpt: TlsOption | null = null,
  ) {
    this.host = host;
    this.port = port;
    this.timeoutMs = timeoutMs;
    this.apiKey = apiKey;
    this.tlsOpt = tlsOpt;
  }

  get connected(): boolean {
    return this.socket !== null && !this.dead;
  }

  connect(): Promise<void> {
    return new Promise<void>((resolve, reject) => {
      const onError = (err: Error): void => {
        this.fail(new ConnectionError(`connect to ${this.host}:${this.port} failed`, { cause: err }));
        reject(this.lastError(err));
      };

      const onReady = (socket: net.Socket): void => {
        socket.removeListener("error", onError);
        socket.setNoDelay(true); // requests are small and latency-sensitive
        this.socket = socket;
        this.lastUsed = Date.now();

        // Start the async reader loop
        this.readLoop(socket).catch((err) => {
          this.fail(new ConnectionError("read loop failed", { cause: err }));
        });

        socket.on("error", (err) => this.fail(new ConnectionError("socket error", { cause: err })));
        socket.on("close", () => this.fail(new ConnectionError("connection closed")));

        if (this.apiKey) {
          this.sessionInit(this.apiKey).then(resolve, reject);
        } else {
          resolve();
        }
      };

      if (this.tlsOpt) {
        const opts: tls.ConnectionOptions =
          this.tlsOpt === true
            ? { host: this.host, port: this.port, servername: this.host }
            : { host: this.host, port: this.port, servername: this.host, ...this.tlsOpt };
        const socket = tls.connect(opts);
        socket.once("error", onError);
        socket.once("secureConnect", () => onReady(socket));
      } else {
        const socket = net.connect({ host: this.host, port: this.port });
        socket.once("error", onError);
        socket.once("connect", () => onReady(socket));
      }
    });
  }

  private lastError(cause: unknown): ConnectionError {
    return new ConnectionError("connection error", { cause });
  }

  private async sessionInit(apiKey: string): Promise<void> {
    const res = await this.request(Cmd.SessionInit, Buffer.from(apiKey, "utf8"));
    if (res.status !== Status.Ok) {
      throw fromStatus(res.status, "SessionInit", res.body);
    }
  }

  /** Send one framed request and resolve with the framed response. */
  request(command: number, payload: Buffer = Buffer.alloc(0)): Promise<Response> {
    return new Promise<Response>((resolve, reject) => {
      if (!this.socket || this.dead) {
        reject(new ConnectionError("not connected"));
        return;
      }
      const timer = setTimeout(() => {
        this.fail(new ConnectionError("request timed out"));
      }, this.timeoutMs);

      this.inFlight.push({ resolve, reject, timer });
      const frame = Buffer.concat([encodeHeader(command, payload.length), payload]);

      this.writeQueue.push({ frame, reject });
      this.pumpWrites();
    });
  }

  private async pumpWrites(): Promise<void> {
    if (this.writing || !this.socket || this.dead) return;
    this.writing = true;

    try {
      this.socket.cork();
      while (this.writeQueue.length > 0) {
        if (this.dead || !this.socket) break;
        const task = this.writeQueue.shift()!;

        const canContinue = this.socket.write(task.frame, (err) => {
          if (err) {
            task.reject(new ConnectionError("write failed", { cause: err }));
            this.fail(new ConnectionError("write failed", { cause: err }));
          }
        });

        // Respect backpressure
        if (!canContinue && !this.dead && this.socket) {
          this.socket.uncork();
          await once(this.socket, "drain");
          if (!this.dead && this.socket) this.socket.cork();
        }
      }
      if (!this.dead && this.socket) this.socket.uncork();
    } catch (err) {
      this.fail(new ConnectionError("write loop error", { cause: err }));
    } finally {
      this.writing = false;
    }
  }

  private async readLoop(socket: net.Socket): Promise<void> {
    let inbox = Buffer.alloc(0);

    for await (const chunk of socket) {
      inbox = inbox.length === 0 ? chunk : Buffer.concat([inbox, chunk]);

      while (inbox.length >= HEADER_LEN) {
        const version = inbox.readUInt8(0);
        if (version !== PROTO_VERSION) {
          throw new ConnectionError(`unexpected protocol version 0x${version.toString(16)}`);
        }

        const status = inbox.readUInt8(1);
        const length = inbox.readUInt32BE(2);

        if (inbox.length < HEADER_LEN + length) break; // wait for more chunks

        const body = inbox.subarray(HEADER_LEN, HEADER_LEN + length);
        inbox = inbox.subarray(HEADER_LEN + length);

        const pending = this.inFlight.shift();
        if (!pending) {
          throw new ConnectionError("received response with no pending request");
        }

        clearTimeout(pending.timer);
        this.lastUsed = Date.now();
        pending.resolve({ status, body: Buffer.from(body) });
      }
    }
  }

  /** Tear down the connection and reject any in-flight requests. */
  private fail(err: ConnectionError): void {
    if (this.dead) return;
    this.dead = true;

    const queue = this.inFlight;
    this.inFlight = [];

    for (const pending of queue) {
      clearTimeout(pending.timer);
      pending.reject(err);
    }

    this.close();
  }

  close(): void {
    if (this.socket) {
      this.socket.removeAllListeners();
      this.socket.destroy();
      this.socket = null;
    }
  }

  /** Send a keepalive; resolves true if the server answered OK. */
  async ping(): Promise<boolean> {
    try {
      const res = await this.request(Cmd.Ping);
      return res.status === Status.Ok;
    } catch {
      return false;
    }
  }
}

File: clients/python/zydecodb/_connection.py
"""A single TCP connection to a ZydecoDB server.

One connection is NOT thread-safe; the pool guarantees a connection is only ever
checked out to one caller at a time. Framing and (optional) authentication live
here; pooling, retries, and the typed API live above.
"""

from __future__ import annotations

import socket
import ssl
import struct
import time
from typing import Optional, Tuple, Union

from . import _protocol as proto
from .errors import ConnectionError as ZConnectionError
from .errors import from_status

# True enables system-default TLS; an SSLContext customizes verification/SNI.
TlsOption = Union[bool, ssl.SSLContext]


class Connection:
    def __init__(
        self,
        host: str,
        port: int,
        *,
        timeout: float,
        api_key: Optional[str],
        tls: Optional[TlsOption] = None,
    ):
        self._addr = (host, port)
        self._timeout = timeout
        self._api_key = api_key
        self._tls = tls
        self._sock: Optional[socket.socket] = None
        self.last_used = 0.0

    def connect(self) -> None:
        self.close()
        try:
            sock = socket.create_connection(self._addr, timeout=self._timeout)
            sock.settimeout(self._timeout)
            # Disable Nagle: requests are small and latency-sensitive.
            sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
            if self._tls:
                ctx = (
                    ssl.create_default_context()
                    if self._tls is True
                    else self._tls
                )
                sock = ctx.wrap_socket(sock, server_hostname=self._addr[0])
        except OSError as exc:
            raise ZConnectionError(f"connect to {self._addr[0]}:{self._addr[1]} failed: {exc}")
        self._sock = sock
        self.last_used = time.monotonic()
        if self._api_key:
            self._session_init(self._api_key)

    def _session_init(self, api_key: str) -> None:
        status, payload = self.request(proto.CMD_SESSION_INIT, api_key.encode("utf-8"))
        if status != proto.STATUS_OK:
            raise from_status(status, "SessionInit", payload)

    @property
    def connected(self) -> bool:
        return self._sock is not None

    def close(self) -> None:
        if self._sock is not None:
            try:
                self._sock.close()
            except OSError:
                pass
            self._sock = None

    def request(self, command: int, payload: bytes = b"") -> Tuple[int, bytes]:
        """Send one framed request and read the framed response. Raises
        `ConnectionError` on any transport failure (the pool discards the
        connection; the client may retry idempotent calls on a fresh one)."""
        if self._sock is None:
            raise ZConnectionError("not connected")
        try:
            self._sock.sendall(proto.encode_header(command, len(payload)) + payload)
            status, body = self._recv()
        except (OSError, ZConnectionError) as exc:
            self.close()
            raise ZConnectionError(f"request failed: {exc}")
        self.last_used = time.monotonic()
        return status, body

    def ping(self) -> bool:
        """Send a keepalive Ping. Returns True if the server answered OK."""
        try:
            status, _ = self.request(proto.CMD_PING)
        except ZConnectionError:
            return False
        return status == proto.STATUS_OK

    def _recv(self) -> Tuple[int, bytes]:
        header = self._recv_exact(proto.HEADER_LEN)
        version, status, length = struct.unpack(">BBI", header)
        if version != proto.PROTO_VERSION:
            raise ZConnectionError(f"unexpected protocol version 0x{version:02x}")
        payload = self._recv_exact(length) if length else b""
        return status, payload

    def _recv_exact(self, n: int) -> bytes:
        assert self._sock is not None
        chunks = []
        got = 0
        while got < n:
            chunk = self._sock.recv(n - got)
            if not chunk:
                raise ZConnectionError("connection closed while reading")
            chunks.append(chunk)
            got += len(chunk)
        return b"".join(chunks)
